tb_plugin/torch_tb_profiler/profiler/memory_parser.py

```diff
@@ -0,0 +1,317 @@
+                    tree_height = len(traverse_dict)
+
+                if current_node is None:
+                    # 3. Ignore all remaining records.
+                    logger.debug("could not find the node for tid %d, timestamp: %d, record index: %d, total records: %d" % (record.tid, record.ts, record_index, len(records)))
+                    self.staled_records.append(records[record_index])
+                    record_index += 1
+                    continue
+
+                if record.ts < current_node.start_time:
+                    # this should only happens for root node.
+                    logger.debug("record timestamp %d is less that the start time of %s" % (record.ts, current_node.name))
+                    # This record has no chance to be appended to following tree node.
+                    self.staled_records.append(record)
+                    record_index += 1
+                    continue
+                elif record.ts >= current_node.end_time:
+                    # 2. pop parent node and update the child_index accordingly.
+                    current_node = current_node.parent
+                    # next child
+                    if current_node is not None:
+                        # pop out the index of parent children. It is current node's index and plus 1 
+                        # So we can continue next node.
+                        child_index = traverse_dict.pop(current_node) + 1
+                    continue
+
+                # 1. find the real node embrace the record.
+                # Find the node which contains the records from top to downmost.
+                # simulate the do while loop as described
+                # https://stackoverflow.com/questions/743164/how-to-emulate-a-do-while-loop
+                while True:
+                    child_find = None
+                    for i in range(child_index, len(current_node.children)):
+                        if record.ts < current_node.children[i].start_time:
+                            # if current record timestamp is less than the current child's startime,
+                            # we will break the search and keep the child_index not change. So that next time 
+                            # we can continue from here.
+                            # there is no any child contains the record.timestamp
+                            break
+                        elif record.ts < current_node.children[i].end_time:
+                            if BENCHMARK_MEMORY:
+                                self.processed_node[current_node.children[i]] += 1
+
+                            traverse_dict[current_node] = i
+                            current_node = current_node.children[i]
+                            child_index = 0
+
+                            # remember the child found for do-while loop
+                            child_find = current_node
+                            break
+
```

Comment:
It's better to add a comment "if record.ts >= current_node.children[i].end_time, continue"